src/deepMerge.ts

```diff
@@ -0,0 +1,68 @@
+import type { MergeDeep } from 'type-fest';
+import { uniq } from './uniq';
+import { isCyclic } from './_isCyclic';
+import { purry } from './purry';
+
+/**
+ * Recursively merges two values, `a` and `b`.
+ * @param target - The first value to merge. This is the dominant parameter in the merge operation.
+ * @param source - The second value to merge.
+ * @signature
+ *    R.deepMerge(a, b)
+ * @example
+ *    R.deepMerge({ foo: 'bar', x: 1 }, { foo: 'baz', y: 2 }) // => { foo: 'bar', x: 1, y: 2 }
+ * @data_first
+ * @category Object
+ * @pipeable
+ */
+export function deepMerge<Target, Source>(
+  target: Target,
+  source: Source
+): MergeDeep<Target, Source>;
+
+/**
+ * Recursively merges two values, `a` and `b`.
+ * @param source - The second value to merge.
+ * @signature
+ *    R.deepMerge(b)(a)
+ * @example
+ *    R.deepMerge({ foo: 'baz', y: 2 })({ foo: 'bar', x: 1 }) // => { foo: 'bar', x: 1, y: 2 }
+ *    R.pipe(
+ *      { foo: 'bar', x: 1 },
+ *      R.difference({ foo: 'baz', y: 2 }),
+ *    ) // => { foo: 'bar', x: 1, y: 2 }
+ * @data_last
+ * @category Object
+ * @pipeable
+ */
+export function deepMerge<Target, Source>(
+  source: Source
+): (target: Target) => MergeDeep<Target, Source>;
+
+export function deepMerge() {
+  return purry(_deepMerge, arguments);
+}
+
+export function _deepMerge(a: unknown, b: unknown): unknown {
+  if (
+    typeof a !== 'object' ||
+    typeof b !== 'object' ||
+    a === null ||
+    b === null
+  ) {
+    return a ?? b;
+  }
+  if (Array.isArray(a)) {
+    if (Array.isArray(b) && !isCyclic(a)) {
+      return a.concat(b);
+    }
+    return a;
+  }
+  const output = { ...a };
```

Comment:
when is it really necessary? it seems that the values for all keys will be overwritten during the loop. I think you can go the other way:

1. easy way: use empty object for output.
```ts
const output = {};
```

2. another way: copy all values through the spread, and let only the shared keys go into the loop.
```ts
const output = { ...a, ...b }; // order is not important 
const keys = intersection(Object.keys(a), Object.keys(b));
for (const k of keys) {
  // @ts-expect-error We've already ensured that "a" and "b" are objects at runtime.
  if (!isCyclic(a[k])) output[k] = deepMerge(a[k], b[k]);
}
return output;
```

3. yet another way: skip keys that are not in `b`: this means that they are only in `a`, which means they are already in the output:

```ts
for (const k of keys) {
  if (!(k in b)) {
    continue;
  }
  // @ts-expect-error We've already ensured that "a" and "b" are objects at runtime.
  if (!isCyclic(a[k])) output[k] = deepMerge(a[k], b[k]);
}
```

in the second and third options, we will save on the number of calls to the `deepMerge` function

Reply:
Chose the second option